\documentclass[12pt,a4paper]{scrartcl}
\usepackage[utf8]{inputenc}
\usepackage{amsmath}
\usepackage{graphicx}
\usepackage{tikz}
%\usepackage{silence}
\usepackage{mdframed}
%\WarningFilter{mdframed}{You got a bad break}
\usepackage[colorinlistoftodos]{todonotes}
\usepackage{listings}
\usepackage{color}
\colorlet{exampcol}{blue!10}
\usepackage{multicol}
\usepackage{booktabs}

\usepackage{exercise}

\usepackage[autostyle, english = american]{csquotes}
\MakeOuterQuote{"}

\usepackage{hyperref}
\hypersetup{
    colorlinks,
    citecolor=black,
    filecolor=black,
    linkcolor=blue,
    urlcolor=black
}

\title{Exercises for statistical inference and stuff}
\date{\today}
\author{Timoth\'ee Bonnet}


\begin{document}


\maketitle

\tableofcontents
\ListOfExerciseInToc
\ExerciseLevelInToc{subsubsection}

\clearpage

\section{Understanding error structure}

\subsection{Broken experiments}

\begin{Exercise}[difficulty=1, title={A significant result?}]
We carry out a vaccine challenge experiment. There are three experimental groups (saline control / low vaccine dose / high vaccine dose), with 6 mice in each group. For each group all mice are housed together in the same cage (so there are three cages). All mice are challenged with \emph{Shigella}, their symptom intensity is scored on day 8. A one-way ANOVA identifies an effect of the treatment, and pair-wise Bonferroni tests show the low-dose to be statistically significant different from the two other groups:

\includegraphics[width=0.4\textwidth]{Figures/message1}

\textbf{What is suspicious in the result? What part of the design may explain the result? How to improve the design (there are at least two different ways)?}
\end{Exercise}


\begin{Exercise}[difficulty=1, title={A better experimental design}]
We study how a membrane protein intakes external molecules in frog eggs. The target molecules are radioactively labelled so we can measure intake. We created five mutant lines to test what part of the protein controls intake. We propose to measure six tubes of ten eggs every week. Each week we will test a different genotype (i.e., the control or one of the five mutants):

\includegraphics[width=0.8\textwidth]{Figures/expdes2}

\textbf{How much information about mutants can we extract from this experiment? How to improve the design?}

\end{Exercise}


\subsection{Broken models?}

Let's look at models with data from the an experiment similar to the mice example.
\begin{Exercise}[difficulty=1, title={lm vs. lmer vs. lm+}]

Load the dataset "challenge2.csv", and test the effect of the vaccine treatment ("Group") on the immune response ("Percent.loss"). 
First fit a fixed-effect only linear model and/or an anova. Is there a clear effect of treatment?\\
Now add a random effect for Cage (e.g., using the package lme4), what happens to the effect of treatment?\\
Finally, fit Cage and Group as fixed effects in a simple linear model and/or anova. What happens, why?\\
Which of the three models should you trust or not?

\end{Exercise}

<<echo=FALSE, eval=FALSE>>=
mice <- read.csv("Data/challenge2.csv")
library(lme4)
summary(lm(Percent.loss ~ 1 + Group, data=mice))
anova(lm(Percent.loss ~ 1 + Group, data=mice))

summary(lmer(Percent.loss ~ 1 + Group + (1|Cage), data=mice))
anova(lmer(Percent.loss ~ 1 + Group + (1|Cage), data=mice))

summary(lm(Percent.loss ~ 1 + Group + Cage, data=mice))
anova(lm(Percent.loss ~ 1 + Group + Cage, data=mice))
@


\section{Interpreting simple mixed models}

\subsection{Summary}
<<echo=FALSE, eval=FALSE>>=
library(lme4)
set.seed(123)
nyear <- 30
nloc <- 20
nbyobs <- 5
nobs <- nyear*nloc*nbyobs
x <- rnorm(nobs)
loc <- sample(1:nloc, size = nobs, replace = TRUE)
years <- sample(1:nyear, size = nobs, replace = TRUE)
y <- 2 + x + rnorm(nloc,0,sqrt(0.2))[loc] + rnorm(nyear,0,sqrt(2))[years] + rnorm(nobs,0,3.1)

dt <- data.frame(x = x, y =y, location=loc, year=years)

write.csv(dt, "Data/smm.csv", row.names = FALSE, quote = FALSE)

summary(lmer(y ~ 1 + x + (1|years), data=dt))
summary(lmer(y ~ 1 + x + (1|loc), data=dt))
summary(lmer(y ~ 1 + x + (1|loc) + (1|years), data=dt))
@

<<echo=FALSE, eval=FALSE>>=
smm <- read.csv("Data/smm.csv")
summary(lmm0 <- lmer(y ~ 1 + x + (1|location) + (1|year), data=smm))
VarCorr(lmm0)
fixef(lmm0)
ranef(lmm0)
sigma(lmm0)

plot(lmm0)
library(lme4)
plot( predict( lmm0, re.form=~0 ), x=smm$x)
plot( predict( lmm0, re.form=~(1|location) ), x=smm$x)
plot( predict( lmm0, re.form=~(1|year) ), x=smm$x)

predict( lmm0, re.form=~0,)

@

\begin{Exercise}[difficulty=1, title={Reading a summary in lme4}]
Load the dataset "smm.csv" to fit a linear mixed model of y as a function of x, with years as a random intercept. 
Does x have a significant effect on y?
How much variation is explained by differences among years? Is there a measure of uncertainty for this estimate?
\end{Exercise}

\begin{Exercise}[difficulty=1, title={Extracting the right number}]
After you fit a linear mixed model in lme4, look at the structure of the model (\texttt{str(model)}). It is very complicated, and notoriously dangerous to try and extract numbers from there. Fortunately, \texttt{summary(), VarCorr(), fixef(), sigma() and ranef()} should give you most of what you need. Try to understand what numbers these functions return.
\end{Exercise}

\begin{Exercise}[difficulty=2, title={Prediction}]
Use the function predict to plot model predictions. Because you predict from a mixed model, you have an option called \texttt{re.form} to control what random effects are included in the prediction. What is the meaning of the predictions from re.form=$\sim 0$, $\sim(1|location)$ and $\sim (1|location) + (1|year)$ ?
\end{Exercise}


\subsection{Tests}
Sometimes random effects are part of the experimental design and are in the models only to control for confounding effects. But sometimes we care about their value or their statistical significance.

\begin{Exercise}[difficulty=2, title={Testing variance components in lme4}]
Use the function anova to test the statistical significance of the random effect "years" and "location". What are the p-values?
How are the p-values calculated here and should you trust this calculation?
\end{Exercise}

\begin{Exercise}[difficulty=1, title={CI for variance components in lme4}]
Use the function confint to estimate confidence intervals for the variance components for "year" and "location".
\end{Exercise}

<<eval=FALSE, echo=FALSE>>=
anova(lmer(y ~ 1 + x + (1|year), data=dt), lm(y~1 + x, data=dt))
anova(lmer(y ~ 1 + x + (1|location), data=dt), lm(y~1 + x, data=dt))

confint(object = lmer(y ~ 1 + x + (1|year), data=dt) )
confint(object = lmer(y ~ 1 + x + (1|location), data=dt) )

library(MCMCglmm)
summary(MCMCglmm(y ~ 1 + x, random = ~year, data=dt))
@

\begin{Exercise}[difficulty=2, title={Testing variance components in MCMCglmm}]
lme4 can have difficulties estimating random effect parameters when models get a bit complex. An very powerful alternative, I recommend MCMCglmm. Try and fit a model of y with x as a fixed effect, and year and location as random effects. Use summary and plot to discuss the importance of the random effects.
\end{Exercise}


\section{Flexible variance structures}

\subsection{Crossed or nested?}
<<echo=FALSE, eval=FALSE>>=
respi <- read.csv("Data/respi.csv")

respi$rrarea1 <- respi$rrarea + rep(x = rnorm(4*20, 0, 0.4), each=4)

lmer(rrarea1 ~ 1 + Genotype * Temp + (1|shelter) + (1|plantID), data=respi)
lmer(rrarea1 ~ 1 + Genotype * Temp +  (1|plantID), data=respi)
lmer(rrarea1 ~ 1 + Genotype * Temp + (1|shelter/plantID), data=respi)


@


\begin{Exercise}[difficulty=2, title={Plants and shelters}]
Load the dataset "respi.csv". We are interested in the interactive effect of genotype and temperature on dark respiration (rrarea1). Plants were measured several time in different temperatures, and we need to control for that.
Fit a mixed model to test the interaction temperature-genotype. Does plantID matter? What if you include shelter as an additional random effect?\\
Plants with the same name in different shelters were not the same plants. How to account for that?\\
How do the variance component for plantID change when you account for it? Why?
\end{Exercise}


\begin{Exercise}[difficulty=1, title={PCR plate}]
What can affect results on a PCR plate? What is the difference between accounting for such effects with the two following random structures?\\
\textbf{Crossed:} \texttt{(1|plate) + (1|row) + (1|column)}\\
\textbf{Nested:} \texttt{(1|plate) + (1|plate:row) + (1|plate:column)}

\includegraphics[width=\textwidth]{Figures/nestedplates}
\end{Exercise}

\subsection{Beyond random intercepts}

So far we have considered random effects around intercepts only (that is the meaning of the "1" in the lme4 syntax (1$\mid$ re)). But random effects can be around fixed effects. You may have heard of "random interactions", "random slopes", "random regressions"\dots

<<echo=FALSE, eval=FALSE>>=
modb <- read.table("modbeetles.txt", header = T)
head(modb)


library(lme4)

summary(lmer(Wo ~ 1 + host + (1|IDf), data = modb))

summary(lmer(Wo ~ 1 + host + (1+host|IDf), data = modb))

anob <- modb[,c("IDf", "host", "Wo", "date.mated")]
names(anob) <- c("parent", "environment", "mass", "cage")

anob$parent <- as.numeric(anob$parent)
anob$cage <- as.numeric(anob$cage)
anob$mass <- (anob$mass - 1) *10 +rnorm(nrow(anob), 0, 0.1)
head(anob)

summary(lmer(mass ~ 1 + environment + (1+environment|parent) + (1|cage), data = anob))
summary(lmer(mass ~ 1 + environment + (0+environment|parent) + (1|cage), data = anob))

summary(lmm2 <- lmer(mass ~ 1 + environment + (0+environment|parent) + (1|cage), data = anob))

plot(ranef(lmm2)$parent)

trs <- matrix(c(1,1, 0,1), nrow = 2)
covie <- -0.58*sqrt(1.0830*6.4103)
vcovie <- matrix(c(6.4103, covie, covie, 1.0830), nrow = 2)

newvcov <- trs %*% vcovie %*% t(trs)
newcor <- newvcov[1,2]/sqrt(newvcov[1,1]*newvcov[2,2])

write.csv(anob, "Data/beetles.csv", row.names = FALSE, quote = FALSE)
@



\begin{Exercise}[difficulty=2, title={Beetles: build a model}]
Load the dataset "beetles.csv". It contains (fake) data from an (real) experiment on gene-by-environment interactions. The variable of interest is the mass of beetles born in two different environments, from different parents, and in different cages. Assuming that we can measure genetic varition with parent random effects, we wonder if different genomes respond differently to different environments.
\textbf{Build the model corresponding to this question in lme4.}

(hints: you could start from a lm() of mass modeled by environment, then add random intercepts, and finally a little something more).
\end{Exercise}

\begin{Exercise}[difficulty=2, title={Beetles: look at the model}]
What are the variances related to genetic differences? How are they correlated? Does genetic variation explain a lot of the total variation we observe? Try and draw a representation of genetic variation in the two environments. 
\end{Exercise}

\begin{Exercise}[difficulty=3, title={Beetles: interpret}]
Interpret model outputs (use raw numbers and / or graphes) to answer the following:
Is there evidence for genetic variation? Do the two environment differ in their effects on beetles? \\
Is there evidence for genetic variation in the response to the environment? \\
Does that mean that genomes good at environment 1 are bad at environment 2?
\end{Exercise}

\subsection{Correlated random effects}

If we have time, ask me to talk or give examples of models with genetic or phylogenetic effects, or spatial auto-correlation.



\end{document}
